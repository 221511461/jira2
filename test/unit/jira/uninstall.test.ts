import testTracking from '../../setup/tracking';

describe('Webhook: /events/uninstalled', () => {
  let installation;
  let subscriptions;
  let uninstall;

  beforeEach(async () => {
    const { getHashedKey } = await import('../../../src/models/installation');
    subscriptions = [{
      gitHubInstallationId: 10,
      jiraHost: 'https://test-host.jira.com',
      uninstall: jest.fn().mockName('uninstall').mockResolvedValue(1),
    }];
    installation = {
      id: 19,
      jiraHost: 'https://test-host.jira.com',
      clientKey: getHashedKey('abc123'),
      enabled: true,
      secrets: 'def234',
      sharedSecret: 'ghi345',
      uninstall: jest.fn().mockName('uninstall').mockResolvedValue(installation),
      subscriptions: jest.fn().mockName('subscriptions').mockResolvedValue(subscriptions),
    };

    td.when(models.Subscription.getAllForHost(installation.jiraHost))
      // Allows us to modify subscriptions before it's finally called
      .thenDo(async () => subscriptions);

    uninstall = (await import('../../../src/jira/uninstall')).default;
  });

  it('Existing Installation', async () => {
    await testTracking();
    const req = { log: jest.fn() };
    const res = { locals: { installation }, sendStatus: jest.fn() };
    await uninstall(req, res);
    expect(res.sendStatus).toHaveBeenCalledWith(204);
    expect(installation.uninstall).toHaveBeenCalled();
    expect(subscriptions[0].uninstall).toHaveBeenCalled();
  });

  it('Existing Installation, no Subscriptions', async () => {
    await testTracking();
    const req = { log: jest.fn() };
    const res = { locals: { installation }, sendStatus: jest.fn() };
    subscriptions = [];
    await uninstall(req, res);
    expect(res.sendStatus).toHaveBeenCalledWith(204);
    expect(installation.uninstall).toHaveBeenCalled();
  });
});
