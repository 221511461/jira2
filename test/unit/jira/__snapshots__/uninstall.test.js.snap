// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook: /events/uninstalled Existing Installation 1`] = `"jira.v0.Action"`;

exports[`Webhook: /events/uninstalled Existing Installation 2`] = `
Object {
  "actionSource": "webhook",
  "association": "installation",
  "githubInstallationId": 10,
  "installationId": 19,
  "jiraHostname": "https://test-host.jira.com",
  "type": "destroyed",
}
`;

exports[`Webhook: /events/uninstalled Existing Installation 3`] = `"jira.v0.Action"`;

exports[`Webhook: /events/uninstalled Existing Installation 4`] = `
Object {
  "actionSource": "webhook",
  "association": "subscription",
  "githubInstallationId": 10,
  "installationId": 19,
  "jiraHostname": "https://test-host.jira.com",
  "type": "destroyed",
}
`;

exports[`Webhook: /events/uninstalled Existing Installation, no Subscriptions 1`] = `"jira.v0.Action"`;

exports[`Webhook: /events/uninstalled Existing Installation, no Subscriptions 2`] = `
Object {
  "actionSource": "webhook",
  "association": "installation",
  "githubInstallationId": 10,
  "installationId": 19,
  "jiraHostname": "https://test-host.jira.com",
  "type": "destroyed",
}
`;
